torch/csrc/jit/tensorexpr/external_functions.cpp

```diff
@@ -0,0 +1,123 @@
+#include <torch/csrc/jit/tensorexpr/external_functions.h>
+
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#include <torch/csrc/jit/tensorexpr/external_functions_registry.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+// A helper function to construct a vector of tensors from raw buffer arguments
+std::vector<at::Tensor> constructTensors(
+    int64_t bufs_num,
+    void** buf_data,
+    int64_t* buf_ranks,
+    int64_t* buf_dims,
+    int8_t* buf_dtypes) {
+  std::vector<void*> buf_data_vec;
+  std::vector<std::vector<int64_t>> buf_dims_vec;
+  std::vector<c10::ScalarType> buf_dtypes_vec;
+  int64_t buf_dims_idx = 0;
+  for (int64_t i = 0; i < bufs_num; i++) {
+    buf_data_vec.push_back(buf_data[i]);
+    buf_dims_vec.push_back({});
+    for (int64_t dim = 0; dim < buf_ranks[i]; dim++) {
+      buf_dims_vec[i].push_back(buf_dims[buf_dims_idx++]);
+    }
+    buf_dtypes_vec.push_back(static_cast<c10::ScalarType>(buf_dtypes[i]));
+  }
+
+  std::vector<at::Tensor> tensors;
+  for (size_t i = 0; i < buf_data_vec.size(); i++) {
+    auto options = at::TensorOptions()
+                       .dtype(buf_dtypes_vec[i])
+                       .layout(at::kStrided)
+                       .device(at::kCPU) // TODO: support GPUs too
+                       .requires_grad(false);
+    tensors.emplace_back(
+        at::from_blob(buf_data_vec[i], buf_dims_vec[i], options));
+  }
+  return tensors;
+}
+
+void nnc_aten_conv2d(
+    int64_t bufs_num,
+    void** buf_data,
+    int64_t* buf_ranks,
+    int64_t* buf_dims,
+    int8_t* buf_dtypes,
+    int64_t args_num,
+    int64_t* extra_args) {
+  std::vector<at::Tensor> tensors =
+      constructTensors(bufs_num, buf_data, buf_ranks, buf_dims, buf_dtypes);
+
+  at::Tensor& r = tensors[0];
+  const at::Tensor& x = tensors[1];
+  const at::Tensor& w = tensors[2];
+  if (args_num > 0) {
+    // Check that if the extra arguments are provided, then the bias tensor is
+    // also present
+    TORCH_INTERNAL_ASSERT(args_num == 7 && bufs_num == 4);
+    const at::Tensor& b = tensors[3];
+
+    int64_t strideH = extra_args[0];
+    int64_t strideW = extra_args[1];
+    int64_t paddingH = extra_args[2];
+    int64_t paddingW = extra_args[3];
+    int64_t dilationH = extra_args[4];
+    int64_t dilationW = extra_args[5];
+    int64_t groups = extra_args[6];
+
+    try {
+      r = at::native::conv2d(
+          x,
+          w,
+          b,
+          {strideH, strideW},
+          {paddingH, paddingW},
+          {dilationH, dilationW},
+          groups);
+    } catch (...) {
+    }
+  } else {
+    try {
+      r = at::native::conv2d(x, w);
+    } catch (...) {
```

Comment:
What's the plan for propagating error codes? I assume we could use `extra_args`, technically, but I don't like that solution.

Reply:
We could change the signature to return an exit code, but the problem is that nothing else in our system expects to deal with such exit codes and knows how to raise them further. While it is possible to build all the infra for this, I think it's more practical to just check sizes/attributes of a node before we lower to NNC making sure a situation an exception occurs would never happen in the TE code.